libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +492,180 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+  } catch (const std::logic_error& e) {
+    LOG(log_er_) << e.what();
+    LOG(log_er_) << "DPOS period " << dpos_period << ", voter " << voter_account_addr << " votes count "
+                 << voter_dpos_votes_count << ", total votes count " << dpos_total_votes_count
+                 << ", pbft sortition threshold " << pbft_sortition_threshold;
+
+    return false;
+  }
+
+  return true;
+}
+
+std::vector<std::shared_ptr<Vote>> VoteManager::updateRewardVotes(uint64_t reward_period) {
+  // TODO[issue1699]
+  auto period_data = db_->getPeriodDataRaw(reward_period);
+  if (period_data.size() == 0) {
+    // First period no reward votes
+    return {};
+  }
+
+  SyncBlock reward_period_sync_block(period_data);
+
+  std::unordered_set<vote_hash_t> reward_period_cert_votes_hashes;
+  for (const auto& v : reward_period_sync_block.cert_votes) {
+    reward_period_cert_votes_hashes.insert(v->getHash());
+  }
+
+  bool update = false;
+  {
+    const auto reward_period_pbft_block_hash = reward_period_sync_block.pbft_blk->getBlockHash();
+    std::unique_lock lock(reward_votes_mutex_);
+    if (reward_votes_.first != reward_period_pbft_block_hash) {
+      LOG(log_dg_) << "Reward period pbft block<" << reward_period_pbft_block_hash
+                   << "> does not match reward votes voted value " << reward_votes_.first;
+      return reward_period_sync_block.cert_votes;
+    }
+    for (auto& v : reward_votes_.second) {
+      if (!reward_period_cert_votes_hashes.contains(v.first) && verifyRewardVote(v.second)) {
+        LOG(log_nf_) << "Add new reward vote " << v.first;
+        update = true;
+        reward_period_cert_votes_hashes.insert(v.first);
+        reward_period_sync_block.cert_votes.push_back(std::move(v.second));
+      }
+    }
+    // Clear reward votes table
+    reward_votes_ = {};
+  }
+
+  if (update) {
+    db_->overridePeriodData(reward_period_sync_block);
+  }
+
+  return reward_period_sync_block.cert_votes;
+}
+
+bool VoteManager::checkRewardVotes(const std::shared_ptr<PbftBlock>& pbft_block,
+                                   const std::vector<std::shared_ptr<Vote>>& reward_period_cert_votes) {
+  if (pbft_block->getPeriod() == 1) {
+    // First period no reward votes
+    return true;
+  }
+
+  std::unordered_set<vote_hash_t> reward_period_cert_votes_set;
+  for (const auto& v : reward_period_cert_votes) {
+    reward_period_cert_votes_set.insert(v->getHash());
+  }
+
+  std::vector<vote_hash_t> missing_reward_votes;
+  const auto& reward_votes = pbft_block->getRewardVotes();
+  for (const auto& v : reward_votes) {
+    if (!reward_period_cert_votes_set.contains(v)) {
+      missing_reward_votes.emplace_back(v);
+    }
+  }
+
+  if (!missing_reward_votes.empty()) {
+    std::ostringstream missing_reward_votes_log;
+    missing_reward_votes_log << "Missing reward votes: ";
+    for (const auto& v : missing_reward_votes) {
+      missing_reward_votes_log << "\n" << v.toString();
```

Comment:
use whitespace, not \n as it is not possible to grep such log with nw lines

Reply:
fixed